docs/rfcs/20220209_limit_connections_rate.md

```diff
@@ -0,0 +1,91 @@
+- Feature Name: Ability to limit connection creation rate per shard
+- Status: in-progress
+- Start Date: 2022-02-09
+- Authors: Vadim Plakhtinskiy
+- Issue: 
+
+# Executive Summary
+
+- Set a limit on the rate with which the shard accepts new connections
+- Set a limit on connection creation rate per IP
+
+
+Similar to KIP-612, adapted to Redpanda's thread per core model
+
+## What is being proposed
+
+New configuration properties (For start we choose global config for connection rate):
+- kafka_max_connection_creation_rate (optional integer, default unset)
+- kafka_max_connection_creation_rate_ovverides (string, default "")
+
+
+When a connection is accepted which exceeds one of the configured
+bounds, it is dropped before reading or writing anything to the socket.
+
+
+## Motivation
+
+* Avoid hitting system resource limits (for example CPU)
+* Avoid geting connections storms
+* Allow broker doing useful work instead of handling high connection creation rate
+
+## How
+
+If both configuration properties are unset, then no action is taken: this
+preserves the existing unlimited behaviour.
+
+We will track conection rate on each shard independently (without cross-core communication).
+
+We will store semaphore to count connection in current second. For ovverides we will store hash table
+from ip -> semaphore.
+
+For new connection we will do:
+1. Update semaphore to limit if need it
+2. Run try_wait for semaphore
+3. If we got true we can continue wiht this connection
+4. If we got false we should decline this connection
+
+struct connection_rate {
+    ss::semaphore current_rate
+    ss::lower_clock last_update_time
+}
+
+
+### Handling a configuration change
+
+Changes to the configuration properties affect all outstanding
+connection_rate objects: the `kafka_max_connection_creation_rate` attribute must be recalculated.
+
+## Impact
+
+Small overhead, because we will use ss::lower_clock to compare time
+
+# Guide-level explanation
+
+## How do we teach this?
+???
+
+## Interaction with other features
+
+* Dependency on centralized config for live updates
+
+# Drawbacks
```

Comment:
Probably worth calling out the drawback of per-shard operation -- that the lowest effective rate the user can set is multiplied by the number of shards.